### PR TITLE
Improve autocomplete sorting

### LINKED ISSUE #5035: [Bug] further improvements on autocomplete
improvements on 'Best match' logic for autocomplete

1. Find all properties up to 3 levels that match the expected type
2. Rank the properties with entities relevant higher based on (widgets for action & vice versa + recent visits)
3. Sort the properties by the manual ranks so the ones we know are useless go to the bottom
4. Cut off the top 3

### PR DESCRIPTION
**Pull request by hetunandu**
## Description

- "Best Match" will never be selected by default
- Best match results are filtered based on the text written in the input
- Results other than best match are not unfurled
- Contextual results are shown before best match (currentRow or vars declared in context)
- Improved data type expected checks
  - Trigger fields will get functions best match results
  - Api header and params keys will get best match results
  - When no expected field is encoded, will not show best match
  - Appsmith object is not listed in best match
- Improved sorting of best match
  - Action fields get widgets results and vice versa
  - Max 3 results are shown in best match
- Removed bunch of JS results that were not valid (promise, Function etc)
- Added some JS keywords in autocomplete
- Improved styling on hint styles

Fixes #5035

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/autocomplete-sorting 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 51.97 **(0.11)** | 33.66 **(0.13)** | 30 **(0.06)** | 52.64 **(0.11)**
 :sparkles: :new: | **app/client/src/actions/evaluationActions.ts** | **58.82** | **0** | **0** | **50**
 :green_circle: | app/client/src/components/editorComponents/CodeEditor/codeEditorUtils.ts | 100 **(0)** | 88.89 **(1.39)** | 100 **(0)** | 100 **(0)**
 :red_circle: | app/client/src/components/editorComponents/CodeEditor/commandsHelper.ts | 16.67 **(-0.28)** | 0 **(0)** | 10 **(0)** | 16.36 **(-0.31)**
 :red_circle: | app/client/src/components/editorComponents/CodeEditor/hintHelpers.ts | 78.26 **(-11.74)** | 75 **(-10.71)** | 40 **(-22.5)** | 77.27 **(-12.09)**
 :red_circle: | app/client/src/components/editorComponents/CodeEditor/index.tsx | 67.37 **(-3.17)** | 41.04 **(-1.93)** | 51.28 **(0)** | 69.27 **(-2.5)**
 :green_circle: | app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx | 57.52 **(0.5)** | 41.67 **(0)** | 25 **(0)** | 59.81 **(0.55)**
 :green_circle: | app/client/src/sagas/EvaluationsSaga.ts | 19.63 **(2.52)** | 3.03 **(2.02)** | 6.67 **(3.44)** | 22.58 **(2.58)**
 :sparkles: :new: | **app/client/src/sagas/PostEvaluationSagas.ts** | **20.47** | **0** | **0** | **24.07**
 :green_circle: | app/client/src/utils/autocomplete/EntityDefinitions.ts | 46.67 **(11.67)** | 0 **(0)** | 25 **(0)** | 46.67 **(11.67)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.83 **(5.9)** | 40.25 **(9.37)** | 36.21 **(5.3)** | 55.04 **(5.2)**
 :green_circle: | app/client/src/utils/autocomplete/customTreeTypeDefCreator.ts | 36.36 **(14.93)** | 100 **(100)** | 0 **(0)** | 30 **(11.48)**
 :green_circle: | app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts | 81.67 **(14.6)** | 57.14 **(-4.4)** | 100 **(0)** | 78.85 **(7.18)**
 :sparkles: :new: | **app/client/src/utils/autocomplete/dataTypeSortRules.ts** | **100** | **100** | **100** | **100**
 :red_circle: | app/client/src/workers/evaluationUtils.ts | 86.62 **(-0.27)** | 70.37 **(-1.6)** | 85.71 **(-1.79)** | 86.4 **(-0.33)**</details>

### CODE REVIEW

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

File: app/client/src/utils/autocomplete/dataTypeSortRules.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    "DROP_DOWN_WIDGET.selectedOptionValue",
+    "DATE_PICKER_WIDGET_2.selectedDate",
+    "DATE_PICKER_WIDGET_2.formattedDate",
+    "TABLE_WIDGET.searchText",
+    "TABS_WIDGET.selectedTab",
+    "IFRAME_WIDGET.source",
+    "IFRAME_WIDGET.title",
+    "DROP_DOWN_WIDGET.selectedOptionLabel",
+    "IMAGE_WIDGET.image",
+    "TEXT_WIDGET.text",
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/utils/autocomplete/dataTypeSortRules.ts:43]: Since it is easier to convert strings to numbers, these are still kept here

**Review by ApekshaBhosale — APPROVED** [review]

**Pull request merged**